### ISSUE TITLE
ccache thinks that -Wp,-DX=Y is interchangeable with -DX=Y

### ISSUE THREAD

**Issue by siddhesh** [ccache/ccache#1238]
Original Fedora bug here:
https://bugzilla.redhat.com/show_bug.cgi?id=2160275

### How to reproduce ###

ccache changes the order of arguments, specifically macro define/undefine arguments passed to the compiler, thus potentially changing the compiled output.

```
$ cat > foo.c
#ifndef _FORTIFY_SOURCE
# warning "_FORTIFY_SOURCE not defined"
#else
# warning "_FORTIFY_SOURCE defined"
#endif
unsigned foo;
$ CCACHE_LOGFILE=ccache.log ccache /bin/g++ -O2 -flto=auto -U_FORTIFY_SOURCE -Wp,-U_FORTIFY_SOURCE -Wp,-D_FORTIFY_SOURCE=3 -o foo.s -S foo.c
```

### Actual behavior ###

```foo.c: warning: -D_FORTIFY_SOURCE defined but value is too low
foo.c:2:3: warning: #warning "_FORTIFY_SOURCE is 0" [-Wcpp]
    2 | # warning "_FORTIFY_SOURCE is 0"
      |   ^~~~~~~
```

### Expected behavior ###

```
foo.c:4:3: warning: #warning "_FORTIFY_SOURCE defined" [-Wcpp]
    4 | # warning "_FORTIFY_SOURCE defined"
      |   ^~~~~~~
```

### Environment ###

ccache-4.7.4-1.fc38.x86_64

**Comment by siddhesh** [comment]
So #738 talks about the reordering, but that seems to only be in the context of warnings. Maybe it's the same underlying issue but I filed a separate one just in case it isn't.

**Comment by jrosdahl** [comment]
Thanks for the bug report.

I can't reproduce this with ccache 4.7.4, neither on my local machine nor in a Fedora 38 Docker container. Please run with `CCACHE_LOGFILE` set and post the log file here.

**Comment by siddhesh** [comment]
Also I realized that I flipped the actual and expected results around, fixing that.

**Comment by siddhesh** [comment]
OK, I updated the reproducer to be more precise.  Here's the CCACHE log you requested, hth.

```
[2023-01-18T08:08:42.145355 887760] === CCACHE 4.7.4 STARTED =========================================
[2023-01-18T08:08:42.145415 887760] Configuration file: /home/siddhesh/.config/ccache/ccache.conf
[2023-01-18T08:08:42.145426 887760] System configuration file: /etc/ccache.conf
[2023-01-18T08:08:42.145426 887760] Config: (default) absolute_paths_in_stderr = false
[2023-01-18T08:08:42.145426 887760] Config: (default) base_dir = 
[2023-01-18T08:08:42.145426 887760] Config: (default) cache_dir = /home/siddhesh/.cache/ccache
[2023-01-18T08:08:42.145426 887760] Config: (default) compiler = 
[2023-01-18T08:08:42.145426 887760] Config: (default) compiler_check = mtime
[2023-01-18T08:08:42.145426 887760] Config: (default) compiler_type = auto
[2023-01-18T08:08:42.145426 887760] Config: (default) compression = true
[2023-01-18T08:08:42.145426 887760] Config: (default) compression_level = 0
[2023-01-18T08:08:42.145426 887760] Config: (default) cpp_extension = 
[2023-01-18T08:08:42.145426 887760] Config: (default) debug = false
[2023-01-18T08:08:42.145426 887760] Config: (default) debug_dir = 
[2023-01-18T08:08:42.145426 887760] Config: (default) depend_mode = false
[2023-01-18T08:08:42.145426 887760] Config: (default) direct_mode = true
[2023-01-18T08:08:42.145426 887760] Config: (default) disable = false
[2023-01-18T08:08:42.145426 887760] Config: (default) extra_files_to_hash = 
[2023-01-18T08:08:42.145426 887760] Config: (default) file_clone = false
[2023-01-18T08:08:42.145426 887760] Config: (default) hard_link = false
[2023-01-18T08:08:42.145426 887760] Config: (default) hash_dir = true
[2023-01-18T08:08:42.145426 887760] Config: (default) ignore_headers_in_manifest = 
[2023-01-18T08:08:42.145426 887760] Config: (default) ignore_options = 
[2023-01-18T08:08:42.145426 887760] Config: (default) inode_cache = true
[2023-01-18T08:08:42.145426 887760] Config: (default) keep_comments_cpp = false
[2023-01-18T08:08:42.145426 887760] Config: (default) limit_multiple = 0.8
[2023-01-18T08:08:42.145426 887760] Config: (environment) log_file = ccache.log
[2023-01-18T08:08:42.145426 887760] Config: (default) max_files = 0
[2023-01-18T08:08:42.145426 887760] Config: (default) max_size = 5.0G
[2023-01-18T08:08:42.145426 887760] Config: (default) msvc_dep_prefix = Note: including file:
[2023-01-18T08:08:42.145426 887760] Config: (default) namespace = 
[2023-01-18T08:08:42.145426 887760] Config: (default) path = 
[2023-01-18T08:08:42.145426 887760] Config: (default) pch_external_checksum = false
[2023-01-18T08:08:42.145426 887760] Config: (default) prefix_command = 
[2023-01-18T08:08:42.145426 887760] Config: (default) prefix_command_cpp = 
[2023-01-18T08:08:42.145426 887760] Config: (default) read_only = false
[2023-01-18T08:08:42.145426 887760] Config: (default) read_only_direct = false
[2023-01-18T08:08:42.145426 887760] Config: (default) recache = false
[2023-01-18T08:08:42.145426 887760] Config: (default) remote_only = false
[2023-01-18T08:08:42.145426 887760] Config: (default) remote_storage = 
[2023-01-18T08:08:42.145426 887760] Config: (default) reshare = false
[2023-01-18T08:08:42.145426 887760] Config: (default) run_second_cpp = true
[2023-01-18T08:08:42.145426 887760] Config: (default) sloppiness = 
[2023-01-18T08:08:42.145426 887760] Config: (default) stats = true
[2023-01-18T08:08:42.145426 887760] Config: (default) stats_log = 
[2023-01-18T08:08:42.145426 887760] Config: (default) temporary_dir = /run/user/1000/ccache-tmp
[2023-01-18T08:08:42.145426 887760] Config: (default) umask = 
[2023-01-18T08:08:42.145552 887760] Command line: ccache /bin/g++ -O2 -flto=auto -U_FORTIFY_SOURCE -Wp,-U_FORTIFY_SOURCE -Wp,-D_FORTIFY_SOURCE=3 -o foo.s -S foo.c
[2023-01-18T08:08:42.145563 887760] Hostname: fedora
[2023-01-18T08:08:42.145568 887760] Working directory: /tmp
[2023-01-18T08:08:42.145572 887760] Compiler type: gcc
[2023-01-18T08:08:42.145632 887760] Unsupported compiler option for direct mode: -Wp,-U_FORTIFY_SOURCE
[2023-01-18T08:08:42.145706 887760] Source file: foo.c
[2023-01-18T08:08:42.145714 887760] Object file: foo.s
[2023-01-18T08:08:42.145861 887760] Running preprocessor
[2023-01-18T08:08:42.145944 887760] Executing /bin/g++ -O2 -flto=auto -S -U_FORTIFY_SOURCE -Wp,-U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -E -o /run/user/1000/ccache-tmp/cpp_stdout.tmp.gOYmgO.i foo.c
[2023-01-18T08:08:42.167252 887760] Got result key from preprocessor
[2023-01-18T08:08:42.167292 887760] Result key: 81e134ce0krp4gggigdbrj36k7fbkhf0c
[2023-01-18T08:08:42.167352 887760] Retrieved 81e134ce0krp4gggigdbrj36k7fbkhf0c from local storage (/home/siddhesh/.cache/ccache/8/1/e134ce0krp4gggigdbrj36k7fbkhf0cR)
[2023-01-18T08:08:42.167566 887760] Reading embedded entry #0 <stderr> (215 bytes)
[2023-01-18T08:08:42.167596 887760] Reading embedded entry #1 .o (3239 bytes)
[2023-01-18T08:08:42.167602 887760] Writing to foo.s
[2023-01-18T08:08:42.167697 887760] Succeeded getting cached result
[2023-01-18T08:08:42.167720 887760] Result: local_storage_hit
[2023-01-18T08:08:42.167724 887760] Result: local_storage_read_hit
[2023-01-18T08:08:42.167728 887760] Result: preprocessed_cache_hit
[2023-01-18T08:08:42.167747 887760] Acquiring /home/siddhesh/.cache/ccache/d/0/stats.lock
[2023-01-18T08:08:42.168022 887760] Acquired /home/siddhesh/.cache/ccache/d/0/stats.lock
[2023-01-18T08:08:42.168564 887760] Releasing /home/siddhesh/.cache/ccache/d/0/stats.lock
[2023-01-18T08:08:42.168637 887760] Unlink /home/siddhesh/.cache/ccache/d/0/stats.lock
[2023-01-18T08:08:42.168645 887760] Released /home/siddhesh/.cache/ccache/d/0/stats.lock
[2023-01-18T08:08:42.168673 887760] Unlink /run/user/1000/ccache-tmp/stderr.tmp.mh6OUL.tmp
[2023-01-18T08:08:42.168690 887760] Unlink /run/user/1000/ccache-tmp/cpp_stdout.tmp.gOYmgO.i
```

**Comment by jrosdahl** [comment]
OK. The issue is not that ccache reorders any options but (as can be seen in the log) that `-Wp,-DX=Y` is rewritten to `-DX=Y`. This was done in 4a8b98b4462c613442a5debf6c6f6b38365a149a (included ccache 3.3+), apparently under the assumption that `-Wp,-DX=Y` and `-DX=Y` are interchangeable and therefore can share cache entries. Will fix.
